# Process
### EXPLORATORY DATA ANALYSIS
### FEATURE ENGINEERING
### DATA PREPROCESSING
### DATA LOADING
### MODEL BUILDING
### MODEL TRAINING
### MODULARIZATION
### EXPERIMENT TRACKING
### MODEL EVALUATION
### INFERENCE PIPELINE
### API DEPLOYMENT
### DOCKERIZATION AND DCLOUD DEPLYOMENT 


In [1]:
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.4 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/markalford/Desktop/python_environments/pytorchDevelopmentEnv/lib/python3.11/site-packages/ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "/Users/markalford/Desktop/python_environments/pytorchDevelopmentEnv/lib/python3.11/site-packages/traitlets/config/application.py", line 1075, in launch_instance
    app.start()
  File "/Users/markalford/Desktop/python_environments/pyto

### EXPLORATORY DATA ANALYSIS
1. data collection
2. seperate data into features and targets
3. handle missing values and completeness
4. ensure correct data dtypes for analysis and modeling
5. handle univariate analysis of target data
6. handle univariate analysis of predictor data
7. check whither missing data correlates with the target
8. handle bivariate analysis for numerical and categorical data
9. establish a correlation structure & multicollinearity
10. check for outlifers and leverage points


In [4]:
data_frame = pd.read_csv('data/heart_disease.csv')
print(data_frame.shape)
display(data_frame.head())
display(data_frame.sample(5))
print(data_frame.dtypes)


(1025, 14)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
749,58,1,1,125,220,0,1,144,0,0.4,1,4,3,1
747,60,1,0,117,230,1,1,160,1,1.4,2,2,3,0
909,50,1,0,144,200,0,0,126,1,0.9,1,0,3,0
20,60,1,2,140,185,0,0,155,0,3.0,1,0,2,0
70,59,1,0,170,326,0,0,140,1,3.4,0,0,3,0


age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object


## Label Meanings
- age = age: 29-77
- sex = what is the sex of the person: 1 = male, 0 = female
- trestbps = resting blood pressure: 94-200mm Hg
- cp = chest pain type (4 values): 0,1,2,3
- chol = serum cholestoral in mg/dl: 126 - 564 mg/dl
- fbs = fasting blood sugar > 120 mg/dl: 1 = true, 0 = false
- restecg = resting electrocardiographic results (values 0,1,2): 0-2
- thalach = maximum heart rate achieved: 71 - 202
- exang = exercise induced angina: 1 = yes, 0 = no
- oldpeak = ST depression induced by exercise relative to rest: 0-6.2
- slope = the slope of the peak exercise ST segment: 0-2
- ca = number of major vessels (0-3) colored by flourosopy: 0-3
- thal: 0 = normal; 1 = fixed defect; 2 = reversable defect
- The names and social security numbers of the patients were recently - removed from the database, replaced with dummy values.

In [ ]:
# extract target from features
target_values = data_frame['target']
target = 'target'
features = [c for c in data_frame.columns if c != target]
feature_values = data_frame[features]
display(target_values.sample(10))
display(feature_values.sample(10))

677    0
160    0
880    1
700    1
508    0
440    0
590    1
604    1
78     1
788    0
Name: target, dtype: int64

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
978,44,1,0,110,197,0,0,177,0,0.0,2,1,2
292,61,1,2,150,243,1,1,137,1,1.0,1,0,2
134,54,1,1,108,309,0,1,156,0,0.0,2,0,3
540,58,1,0,128,259,0,0,130,1,3.0,1,2,3
730,44,1,1,120,220,0,1,170,0,0.0,2,0,2
225,54,1,2,150,232,0,0,165,0,1.6,2,0,3
802,63,0,0,124,197,0,1,136,1,0.0,1,0,2
258,38,1,3,120,231,0,1,182,1,3.8,1,0,3
803,52,1,0,112,230,0,1,160,0,0.0,2,1,2
418,62,0,0,124,209,0,1,163,0,0.0,2,0,2
